### ISSUE TITLE
If a process crashes half-way through writing a cache entry to a file cache, it will get corrupted

### ISSUE THREAD

**Issue by itamarst** [psf/cachecontrol#325]
For FileCache this may not be a problem insofar as deserializing will fail? For SeparateBodyFileCache it will be, insofar as it's just a pile-of-bytes for the data.

`pip` doesn't have this problem since it writes to a temporary file and then uses `os.replace()`. CacheControl may wish to adopt this.

**Comment by woodruffw** [comment]
> `pip` doesn't have this problem since it writes to a temporary file and then uses `os.replace()`. CacheControl may wish to adopt this.

Makes sense. Would you be interested in sending a patch for this? Otherwise, I'll possibly have some time to look into this in the coming weeks (but no guarantees).

**Comment by itamarst** [comment]
The relevant code in pip is https://github.com/pypa/pip/blob/8a0f77c171d60344e6a3bf6e95ad5740c21575fd/src/pip/_internal/network/cache.py#L57, could probably just copy-paste the relevant utility code and integrate with the locking.

**Comment by Nikhil172913832** [comment]
Thanks for the update. I’ll try to reproduce this on the current version and report back. If I can still reproduce it, I’ll submit a patch.

**Comment by Nikhil172913832** [comment]
Follow-up: I tried reproducing this on the current version but couldn’t reproduce the issue. It looks like it’s already been fixed.
